lib/SharedCUEParser/SharedCUEParser.cpp

```diff
@@ -42,79 +43,79 @@ static void write_default_cuesheet(char * cue_sheet)
 {
-    set("");
+    m_cue_sheet = _shared_cuesheet;
+    switch_cue();
 }
 
 SharedCUEParser::SharedCUEParser(const char* path)
 {
-    set(path);
+    m_cue_sheet = _shared_cuesheet;
+    _cue_file.open(path);
+    switch_cue();
 }
 
- void SharedCUEParser::set(const char* path)
- {
-    strlcpy(_cue_filepath, path, sizeof(_cue_filepath));
-    m_cue_sheet = _shared_cuesheet;
-    if (path[0] == '\0' && _shared_cuesheet[0] == '\0')
-    {
-        write_default_cuesheet(_shared_cuesheet);
-    }
-    restart();
- }
+FsFile *SharedCUEParser::get_cue_file()
+{
+    return &_cue_file;
+}
 
 // Restart parsing from beginning of file
 void SharedCUEParser::restart()
 {
-    update_file();
+    switch_cue();
     CUEParser::restart();
 }
 const CUETrackInfo *SharedCUEParser::next_track()
 { 
-    update_file();
+
+    switch_cue();
     return CUEParser::next_track();
 }
 
 
 const CUETrackInfo *SharedCUEParser::next_track(uint64_t prev_file_size)
 {
-    update_file();
+    switch_cue();
     return CUEParser::next_track(prev_file_size);
 }
 
-void SharedCUEParser::update_file()
+void SharedCUEParser::load_updated_cue()
 {
-    if ( strcasecmp(_cue_filepath, _current_file_loaded) != 0)
+    load_cue();
+    CUEParser::restart();
+}
+
+void SharedCUEParser::load_cue()
+{
+    if (!_cue_file.isOpen())
+    {
+        write_default_cuesheet(_shared_cuesheet);
+    }
+    else
     {
-        strlcpy(_current_file_loaded, _cue_filepath, sizeof(_current_file_loaded));
-        // Empty filepath, load simple cuesheet
-        if (_current_file_loaded[0] == '\0')
+        _cue_file.rewind();
+        int count = _cue_file.read(_shared_cuesheet, sizeof(_shared_cuesheet) - 1);
+        char filename[257];
+        _cue_file.getName(filename, sizeof(filename));
+        // on read error, close _cue_file;
+        if (count <= 0)
         {
-            write_default_cuesheet(_shared_cuesheet);
+            _cue_file.close();
         }
         else
         {
-
-            FsFile file = SD.open(_current_file_loaded);
-            if (file.isOpen())
-            {
-                int count = file.read(_shared_cuesheet, sizeof(_shared_cuesheet));
-                file.close();
-                // on read error, set _shared_cuesheet to an empty string;
-                if (count <= 0)
-                {
-                    _shared_cuesheet[0] = '\0';
-                }
-                else
-                {
-                    // Null terminate data into a valid string
-                    _shared_cuesheet[count] = '\0';
-                }
-            }
-            else
-            {
-                // on open error, set _shared_cuesheet to an empty string;
-                _shared_cuesheet[0] = '\0';
-            }
+            // Null terminate data into a valid string
+            _shared_cuesheet[count] = '\0';
         }
     }
 }
 
+void SharedCUEParser::switch_cue()
+{
+    if ( _current_cue_file != &_cue_file)
```

Comment:
There is a tiny optimization if comparing `_cue_file.firstCluster()`, because the file wouldn't have to be reloaded when ide_cdrom.cpp first reads it and then the audio.cpp. But I agree it is insignificant for actual usage.